core/node/node_sync/src/batch_transaction_updater/mod.rs

```diff
@@ -0,0 +1,302 @@
+//! Component responsible for updating L1 batch status.
+
+use std::time::Duration;
+
+use tokio::sync::watch;
+use zksync_dal::{Connection, ConnectionPool, Core, CoreDal};
+use zksync_eth_client::EthInterface;
+use zksync_health_check::{Health, HealthStatus, HealthUpdater, ReactiveHealthCheck};
+use zksync_types::{
+    aggregated_operations::AggregatedActionType,
+    eth_sender::{EthTxFinalityStatus, L1BlockNumbers, TxHistory},
+    web3::TransactionReceipt,
+    Address, SLChainId,
+};
+
+use self::l1_transaction_verifier::L1TransactionVerifier;
+
+mod l1_transaction_verifier;
+
+#[cfg(test)]
+mod tests;
+
+/// Module responsible for updating L1 batch status. (Pending->FastFinalized->Finalized)
+/// Its currently the component that should terminate EN when gateway migration happens.
+/// The EN only is stopped when all batches from previous SL are finalized.
+#[derive(Debug)]
+pub struct BatchTransactionUpdater {
+    sl_client: Box<dyn EthInterface>,
+    l1_transaction_verifier: L1TransactionVerifier,
+    pool: ConnectionPool<Core>,
+    health_updater: HealthUpdater,
+    sleep_interval: Duration,
+    processing_batch_size: i64,
+}
+
+impl BatchTransactionUpdater {
+    pub fn new(
+        sl_client: Box<dyn EthInterface>,
+        diamond_proxy_addr: Address,
+        pool: ConnectionPool<Core>,
+        sleep_interval: Duration,
+        processing_batch_size: i64,
+    ) -> Self {
+        Self::from_parts(
+            sl_client,
+            diamond_proxy_addr,
+            pool,
+            sleep_interval,
+            processing_batch_size,
+        )
+    }
+
+    fn from_parts(
+        sl_client: Box<dyn EthInterface>,
+        diamond_proxy_addr: Address,
+        pool: ConnectionPool<Core>,
+        sleep_interval: Duration,
+        processing_batch_size: i64,
+    ) -> Self {
+        Self {
+            sl_client,
+            l1_transaction_verifier: L1TransactionVerifier::new(diamond_proxy_addr, pool.clone()),
+            pool,
+            health_updater: ReactiveHealthCheck::new("batch_transaction_updater").1,
+            sleep_interval,
+            processing_batch_size,
+        }
+    }
+
+    pub fn health_check(&self) -> ReactiveHealthCheck {
+        self.health_updater.subscribe()
+    }
+
+    async fn apply_status_update(
+        &self,
+        connection: &mut Connection<'_, Core>,
+        db_eth_history_id: u32,
+        receipt: TransactionReceipt,
+        l1_block_numbers: &L1BlockNumbers,
+    ) -> anyhow::Result<()> {
+        let updated_status =
+            l1_block_numbers.get_finality_status_for_block(receipt.block_number.unwrap().as_u32());
+
+        if updated_status == EthTxFinalityStatus::Pending {
+            // Errors from apply_status_update will result in a warning and retry
+            anyhow::bail!(
+                "Transaction {} is still pending on SL",
+                receipt.transaction_hash
+            );
```

Comment:
I *really* don't like it that *all* errors are retried. It makes sense to distinguish internal errors and validation errors (both are persistent). Internal errors should definitely be fatal; not sure about validation ones.